### DISCUSSION TITLE
Non-React custom tag example

### DISCUSSION THREAD

**Discussion by n-older** [markdoc/markdoc#319] [Q&A] [answered]
Hello. We're building some help docs for our app using Sanity.io as a CMS which contains a bunch of markdown files that we fetch and render into static pages using Middleman as a static site generator. I'm trying to mimic the examples of custom tags from the docs but they all seem to be catered toward someone using React which is no good for our use case.

I am having difficulty finding any sort of example of a custom tag suited for the HTML renderer, or perhaps I'm not understanding the process properly. Any guidance would be much appreciated.

**Comment by rpaul-stripe** [comment, marked as answer]
When using Markdoc's HTML renderer, you can think of custom tag transform functions as a server-side templating system for generating the markup that you want to output. If you have a concrete example of what you want to accomplish with a custom tag, I can likely show you how to do that with a transform function.

**Reply by n-older** [reply]
That would be great, just so I can get an understanding of how the API works. We'd love to have both the tabs and callout like the examples in the docs.

**Reply by rpaul-stripe** [reply]
Here's a very quick and dirty example of a callout and a tab switcher using the HTML renderer: https://gist.github.com/rpaul-stripe/eb70085c04738fbccf11a40cbe30b2f0

You generally need some kind of client-side code for the tab switcher, so in this example I used JQuery UI. You can replace that with whatever you want, but the general idea is that you use the tag transform function to output the HTML markup that your tab switcher logic expects.

**Reply by n-older** [reply]
Perfect. Yeah we were mainly just looking for a way to generate the HTML structure we wanted so we can hook up an existing Stimulus controller to handle the interactivity client side. Looking at your example makes it much clearer what is happening.

That being said, is there a page that goes into more detail on the parameter and argument types of a tag's transform property as well as the `Tag` class? For example in the [docs](https://markdoc.dev/docs/nodes#options):
```javascript
(Ast.Node, ?Options) =>
  | RenderableTreeNode
  | RenderableTreeNode[]
  | null
```
I could not find any type definitions listed for `Ast.Node`, `Options`, or `RenderableTreeNode` so I was a bit unsure of exactly what needs to go in and out. And likewise for the `Tag` class. I could infer based on the examples but not be 100% sure that I'm passing in exactly what I need to.

**Reply by rpaul-stripe** [reply]
We don't really have a lot of API reference documentation for these things yet—I should probably fill in some of those gaps.

The `Tag` is really simple, it takes three parameters: the name of the HTML tag to render, an object with the attributes and corresponding values, and an array of children. The children can be other `Tag` instances or plain strings.

In the `transform` function, the `Options` parameter is the [top-level configuration object](https://markdoc.dev/docs/config) passed into `Markdoc.transform` as the second parameter. The `Ast.Node` class is defined [here](https://github.com/markdoc/markdoc/blob/main/src/ast/node.ts). The main thing that's good to know here is that `transformChildren` will convert the node's children into the `Tag` equivalents.

**Reply by n-older** [reply]
No worries, I know this is a fairly new project and writing documentation is no walk in the park.

Okay great, that is good information to know. Thank you very much for all your help!

**Reply by rpaul-stripe** [reply]
You are welcome! One other suggestion I have been meaning to document is that it can be helpful to use a library like [htm](https://github.com/developit/htm) in order to make transform functions more readable and maintainable. I put together a modified version of the previous example to demonstrate that: https://gist.github.com/rpaul-stripe/4077c66b7ad7b4811927d2fd7172bfac

**Reply by n-older** [reply]
Oh wow that is much more readable. I actually was curious if there was any sort of way to template out the returned HTML much like you can with a JSX component but looks like that solves that mystery. Didn't even know about htm until you mentioned it but I'm sure my team and I will appreciate the readability so thanks again!

**Reply by drusellers** [reply]
Oh, dang. That is waaaay better than what I have been doing. 10/10 to add this to the docs.
